### ISSUE TITLE
Make originId obligatory 

### ISSUE THREAD

**Issue by agluszak** [build-server-protocol/build-server-protocol#457]
Currently only some requests allow to provide an originId, and those which do make it optional. Our proposal is to make originId obligatory for all requests and some notifications (task notifications, publishDiagnostics). That will allow clients to match notifications to requests. 

It's going to affect #456

**Comment by agluszak** [comment]
Also:

IMO originId field in CompileParams, CompileResult,  CompileReport, TestParams, TestResult, RunParams, and RunResult is redundant and also inconsistent with the rest. It is redundant because [`Tasks that are spawned by another task should reference the originating task's taskId in their own taskId's parent field. Tasks spawned directly by a request should reference the request's originId parent.`](https://build-server-protocol.github.io/docs/specification#task-notifications), and TaskId is already part of TaskStartParams, TaskProgressParams and TaskFinishParams. Inconsistent because TestReport, TestTask,  TestStart, TestFinish and DebugSessionParams don't have that field.

My assumption is that clients are expected to track the parent-child task tree structure anyway, and originId is simply always the top-level parent.

**Comment by jastice** [comment]
The protocol deliberately allows for tasks and diagnostics notifications to be not associated with any request. For instance, the build server might support automatic building on file change. In this case, we would like to report tasks and diagnostics without originId

**Comment by adpi2** [comment]
> The protocol deliberately allows for tasks and diagnostics notifications to be not associated with any request. For instance, the build server might support automatic building on file change. In this case, we would like to report tasks and diagnostics without originId
:+1: Also if more than one client connect to the same server they should all receive the notifications triggered by the other clients' request .

**Comment by agluszak** [comment]
@jastice okay, this makes sense. But origin id in requests should still become obligatory. This will allow us to tell the case you described from tasks initiated by the client

**Comment by agluszak** [comment]
@adpi2 what do you mean by "more than one client connect to the same server"? :O

**Comment by adpi2** [comment]
> I think it can be left to the server implementors if they want to allow multiple connections. It does have valid use cases, for example sending the output to both IDE and terminal, and sbt supports this already.

Agreed and it is not specific to sbt. Also Mill and Bloop support multiple connections and we use them quite a lot. The main use-case is to connect simultaneously from the terminal and the IDE, and it allows us to avoid compile-time duplication. Once your code has compiled from your IDE you can run it, or run the tests, immediately from your terminal or vice-versa, you don't need to wait for a second compilation. Another use-case is to open your project in two different IDEs, that's less common but still useful.

> having multiple clients could mean the servers have to address a range of new problems typical to distributed computing,

That's true and I think all build tools have to deal with this in one way or another. Either they synchronize concurrent compilations or they isolate them in a safe environments.

> @jastice okay, this makes sense. But origin id in requests should still become obligatory. This will allow us to tell the case you described from tasks initiated by the client

That's tricky because the requestId comes from the client and two clients can use the same requestId. The server can deal with this and decide whether to set the `originId` or not based on the client it is sending the notification to. But that implies the `originId` should be optional.

**Comment by agluszak** [comment]
> Also Mill and Bloop support multiple connections and we use them quite a lot. The main use-case is to connect simultaneously from the terminal and the IDE, and it allows us to avoid compile-time duplication. Once your code has compiled from your IDE you can run it, or run the tests, immediately from your terminal, avoiding a second compilation. Another use-case is to open your project in two different IDEs, that's less common but still useful.

Why not introduce some kind of multiplexer/intermediary which would gather requests from multiple clients and forward them to the server? It could also distribute notifications from the server to everyone interested.

> That's tricky because the requestId comes from the client and two clients can use the same requestId. The server can deal with this and decide whether to set the originId or not based on the client it is sending the notification to. But that implies the originId should be optional.

I don't understand what you mean and don't see the implication you're talking about.

Two clients can use the same originId, yes. But a single client can also in theory use the same originId twice. In both cases using the same originId twice should be treated as error.
(Again, using the multiplexer/intermediary would solve this problem, because it could take care of assigning originIds).

**Comment by adpi2** [comment]
> Why not introduce some kind of multiplexer/intermediary which would gather requests from multiple clients and forward them to the server? It could also distribute notifications from the server to everyone interested.

The sbt, Mill and Bloop server are already able to receive requests from multiple clients and to distribute notifications to all of them. My point is, multiple connections is allowed and should stay allowed.

> I don't understand what you mean and don't see the implication you're talking about.

Maybe we need to clarify a few things, because I am not even sure to understand the purpose of this issu.
- In JSON RPC, all requests have a request `id` and it is not optional. So why would we like to add an `originId` into a request ?
- `originId` is useful in a notification, to know which request triggered it. In other words, `originId` in a notification contains the `id` of the request that triggered it.
- Should `originId` not be optional in a notifiaction? In general I think it should be optional because a notification can be triggered by some event that is not a BSP request.
